pysb/simulator/scipyode.py

```diff
@@ -551,6 +417,353 @@ def _integrator_process(initials, param_values, code_eqs, jac_eqs, num_species,
+    def _get_jacobian(self):
+        """Return the Jacobian function jac(time, y, params, const_exprs).
+
+        Subclasses may implement this method for Jacobian support."""
+        raise NotImplementedError
+
+
+class PythonRhsBuilder(RhsBuilder):
+
+    def _get_rhs(self):
+        kinetics = sympy.lambdify(
+            [self.y, self.p, self.e, self.o], self.kinetics
+        )
+
+        def rhs(t, y, p, e):
+            o = (self.observables_matrix * y)[:, None]
+            v = kinetics(y[:, None], p[:, None], e, o)
+            ydot = self.stoichiometry_matrix * v[:, 0]
+            return ydot
+
+        return rhs
+
+    def _get_jacobian(self):
+        kinetics_jacobian_y = sympy.lambdify(
+            [self.y, self.p, self.e, self.o], self.kinetics_jacobian_y
+        )
+        kinetics_jacobian_o = sympy.lambdify(
+            [self.y, self.p, self.e, self.o], self.kinetics_jacobian_o
+        )
+
+        def jacobian(t, y, p, e):
+            o = (self.observables_matrix * y)[:, None]
+            dy = kinetics_jacobian_y(y[:, None], p[:, None], e, o)
+            do = kinetics_jacobian_o(y[:, None], p[:, None], e, o)
+            # Compute Jacobian of the kinetics vector by total derivative.
+            jv = dy + do * self.observables_matrix
+            jac = (self.stoichiometry_matrix * jv).todense()
+            return jac
+
+        return jacobian
+
+
+def _mat_sym_dims(symbol):
+    """Return codegen Argument dimensions for a MatrixSymbol."""
+    return ((0, symbol.shape[0] - 1), (0, symbol.shape[1] - 1))
+
+
+class CythonRhsBuilder(RhsBuilder):
+
+    def __init__(self, model, with_jacobian=False, cleanup=True):
+        super(CythonRhsBuilder, self).__init__(model, with_jacobian, cleanup)
+        routine_names = ["kinetics"]
+        if with_jacobian:
+            routine_names += ["kinetics_jacobian_y", "kinetics_jacobian_o"]
+        # We want more control over various details of code generation and
+        # wrapper module creation than sympy's autowrap provides, so we'll use
+        # the lower-level building blocks directly.
+        routines = {name: self._build_routine(name) for name in routine_names}
+        code_gen = C99CodeGen()
+        code_wrapper = CythonCodeWrapper(
+            code_gen,
+            filepath=self.work_path,
+            extra_compile_args=[
+                # The RHS evaluation is such a tiny part of overall integration
+                # time, even for huge models, that compiler optimization
+                # actually takes more time than it will ever yield back. Since
+                # Cython sets -O2 by default we need to override it.
+                "-O0",
+                # TODO: Enable this line once Cython 3.0 is released.
+                #"-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION",
+            ],
+        )
+        escaped_name = re.sub(
+            r"[^A-Za-z0-9]",
+            "_",
+            self.model_name.encode("unicode_escape").decode(),
+        )
```

Comment:
I guess there might be some edge cases where this causes a mapping collision, but since a unique temp directory is used this shouldn't cause any issues.